### DISCUSSION TITLE
Implement Shadow Part

### DISCUSSION THREAD

**Discussion by jorgecasar** [ing-bank/lion#1137] [Ideas]
Could be interesting implement Shadow Parts to be able style Shadow DOM.

We could discuss about part names and which parts we would like to expose.

First approach I suggest is to declare parts with the end of the class name, the part after "__".

For example in the FormControl would be something like:

```
<div part="group-one" class="form-field__group-one">
  <div part="label" class="form-field__label">
    <slot name="label"></slot>
  </div>
  <small part="help-text" class="form-field__help-text">
    <slot name="help-text"></slot>
  </small>
</div>
<div part="group-two" class="form-field__group-two">
  <div class="input-group">
    <div part="before" class="input-group__before">
      <slot name="before"></slot>
    </div>
    <div part="container" class="input-group__container">
      <div part="prefix" class="input-group__prefix">
        <slot name="prefix"></slot>
      </div>
      <div part="input" class="input-group__input">
        <slot name="input"></slot>
      </div>
      <div part="suffix" class="input-group__suffix">
        <slot name="suffix"></slot>
      </div>
    </div>
    <div part="after" class="input-group__after">
      <slot name="after"></slot>
    </div>
  </div>
  <div part="feedback" class="form-field__feedback">
    <slot name="feedback"></slot>
  </div>
</div>
```

Related PR: https://github.com/ing-bank/lion/pull/1138

**Comment by jorenbroekema** [comment]
@jdelight Brought up some good points in the PR. I think the 3 ways he outlined are correct indeed. To summarize:

1. Extend component. Overwrite `static get styles` / `render` methods where needed. Use case: Design System (DS) built on top of lion. Disadvantage: more maintenance in the DS, and with overriding `render`, could mean that you might miss some updates from lion side e.g. when some extra semantic html is added for a11y for example. Less likely with styles as CSS cascades and you'd always include the super.styles, with render this is not always practical..
2. CSS Parts. Style exposed parts directly without the need to extend. Use case: Using Lion elements directly, but also in a DS it saves having to create an extension, when a wrapper with some CSS to address the parts might be enough. Disadvantage: requires DS to unexpose parts if they don't want their app/feature developers to be able to use the parts to make variations that are not intended by the DS. Need a POC on this to see how unexposing parts would work, @jdelight suggests also looking at [exportparts](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes#attr-exportparts).
3. CSS custom properties. See recent [Chrome dev summit](https://www.youtube.com/watch?t=1042). Use case: only expose CSS rules that you want to expose, and have stricter Design Compliance in your DS. Disadvantage: more CSS custom props needed than the amount of parts that would be needed to create the same flexibility.

I agree with @jdelight that 3 seems like the best way for design systems, and this will always be the main use case for Lion. Yes you'll need more CSS custom props, but parts seems like more maintenance and tricky to get strict design compliance for our users that build design systems on top of lion.

**Reply by jorenbroekema** [reply]
I talked a bit with some of the team on WhatsApp and we'll definitely continue discussion after the holidays. We do agree that parts would be great for the use case of using lion directly with just CSS, and that's even a possibility for simple design systems.

For unexposing CSS parts in a design system I want to add that there is one easy way to unexpose parts, and one that requires a little more thought / implementation in lion.

1. If the basic functional styling still uses normal classes, but parts are exposed on top (so CSS in lion remains the same but the HTML/template now contains parts), the way to "unexpose" in the design system layer would be to just override the HTML/template and not include those parts. That's easy enough! But the disadvantage is that you end up maintaining / overriding the HTML/template of a component only for the sake of unexposing the parts. That's not very nice.
2. We could add a helper inside Lion that allows for easy re-exposing of parts.
  Pseudo code/initial API proposal, a whitelist of the parts you want (re-)exposed in your design system:
    ```js
    static get exposeParts() {
      return ['label', 'input', 'feedback']; // default is [] for extensions, in LionBase component itself, it considers all parts as exposed
    }
    ```
    That way, if you extend a component from lion, you don't get the parts unless you opt-in to having them exposed in your layer. That way exposing a part in an extension is a conscious choice and not a potentially unhappy accident :).
    Another way would be a blacklist, in case we cannot determine properly whether a class is a Lion class or an extension class.
    Then it becomes like:
    ```js
    static get unexposeParts() {
      return ['label', 'input', 'feedback']; // default is [] meaning keep all exposed
    }
    ```
    That would be a bit easier implementation wise but I think a whitelist is better than a blacklist for our users, so they consciously opt into which parts they want to expose.
    For both suggestions we can make a shortcut value to expose/unexpose "everything" e.g. `true` or `'all'` or something like that.

**Reply by jorgecasar** [reply]
Attribute `exportparts` is used to expose parts of components that are inside Shadow DOM of other components. I guess this case doesn't exist in Lion. Maybe you have more visibility about that.

About options 3. To provide enough flexibility I guess we have to add a lot of CSS custom properties and couldn't be manage easily by Lion Team, neither final user. You can take a look how many css I had to change to style calendar (https://github.com/ing-bank/lion/pull/1138/files#diff-75af82f0d61fd86d3bdbdc9988c45825d066d3435d84493c6322244b849f5fca). Can you imagine move this approach to CSS custom properties? It could be more than 100 CSS custom properties.

**Reply by jorenbroekema** [reply]
Well yeah it's 1 CSS prop for every design token that's adjustable by your user. The stricter your design system, the fewer CSS custom props you use and the more fixed values you set. So for design systems which allow a ton of flexibility, it would get pretty crazy indeed :P and for those I definitely agree parts are a far better experience. I think with the proposal in my reply above we can get best of both worlds, where strict design systems can just not re-expose any of the parts, and flexible design systems can expose them to allow users to go more wild :)

Edit: ah I read a bit more into exportparts, I don't think it's very applicable here indeed

**Reply by jorgecasar** [reply]
The 2nd implementation looks nice, then you can use lion components and once you extend them you can decide which parts expose. We should have to define the return cases. If I understood well:

1. empty array (default): expose all
2. null: doens't expose any part.
3. other: expose parts in array

**Reply by jorenbroekema** [reply]
Ehh almost, the default would be to not expose any parts if you extend e.g. `class WoofInput extends LionInput`. But if you use `<lion-input>` directly, it will have all the parts exposed. So for direct users, get every part. For extending users, opt-in. Perhaps allow to not just return array but also `true` or `'all'` if you want a shortcut for opting into all parts.

**Reply by jorgecasar** [reply]
Ummm, I not sure how to detect if the class is a Lion or extends from it and decide to check `exposeParts` or not.

Are you thinking in something like this?

```javascript
class LionPartExample extends LitElement {
   static get exposeParts() {
      return this.name === 'LionPartExample' ? ['one', 'two'] : [];
   }
   
   render() {
      return html`
         <p${this.renderPartOne()}></p>
         <p${this.renderPartTwo()}></p>
      `;
   }
   
   static checkexposePart(name) {
      return  this.exposeParts === 'all' || name in this.exposeParts();
   }
   
   renderPartOne() {
      return this.constructor.checkexposePart('one') ? html` part="one"` : '';
   }
   
   renderPartTwo() {
      return this.constructor.checkexposePart('one') ? html` part="two"` : '';
   }
}
```

Without override:

```javascript
class DefaultPartExample extends LionPartExample { }
DefaultPartExample.exposeParts(); // []
```

Override `exposeParts`;

```javascript
class OnePartExample extends LionPartExample {
   static get exposeParts() {
      return ['one'];
   }
}
DefaultPartExample.exposeParts(); // ['one']
```

Using part you can access to the node attributes, then you have to expose all attributes as part to be able to style it. For example, in the calendar, we have to expose "today", "disabled", "selected", etc... Here you have the full example:

https://github.com/ing-bank/lion/pull/1138/files#diff-878ed2b6426fafeb430a5c460dc4a1f65d2901a98b33d2f2e77d2cdddd3b7565

**Reply by jorenbroekema** [reply]
Well I hadn't thought that far ahead yet on how to detect whether a class is a Lion class or an extension, but yes the idea would be to have something long the lines of like:

```js
class LionField {
  static get exposeParts() {
      return this.isLionBaseClass() ? 'all' : [];
   }
}
```

I'm thinking LionField but perhaps we need a higher level class, perhaps our re-export of LitElement itself.. so the logic applies to all of lion elements. Not sure...

And then yes like you suggested, the rendering of the part would be conditional on whether the name of the part is included in the array or if the exposeParts returns the shortcut for all parts (`'all'` or `true` or similar).

A naive implementation of `isLionBaseClass()` could be `this.constructor.name.startsWith('Lion')` but the problem is that bundlers, transpilers, compilers, don't support this because when you do the transpilation/compilation, very often the name of the class is changed or removed :\. So we need a smarter way there, not sure if there is one... if it's not possible we can do the blacklist way, so all parts always exposed (starting at lion), and in design systems you can opt-out. So we just reverse the proposal in that way 🙈

**Reply by jorenbroekema** [reply]
I discussed this with the team a bit. Seems like we're in favor if we can do the whitelist approach where by default parts are exposed on Lion base classes, but unexposed by default on extension classes, and you have to use the "parts exposure API" to address which parts you want to expose in your extension. It's a bit tricky with regards to technical implementation though.. will need to think of something smart.

___

If the above is technically not possible, we can do the blacklist approach where all parts are exposed. 

 **Disclaimer:** There is no consensus yet in the team that this would be an "acceptable" solution, since there were some objections.. so take with a grain of salt and just food for discussion.

If, on the extension, you don't want any parts exposed, you can just write a mixin:

```js
const NoPartsMixin = (superclass = class {}) => class extends superclass {
  static get exposedParts() { 
    return [];
  }
};

class WoofInput extends NoPartsMixin(LionInput) {}
```

Maybe there's even a better API for this but it's the best I saw in our team chat :P feel free to suggest something better. Keep in mind our design system extension inside ING, and probably many others, are strict. This means they need parts to not be exposed, because they don't want their users to be able to "abuse" parts to create variations that are not design-compliant.

**Reply by jorgecasar** [reply]
I created a [`ShadowPartsMixin`](https://github.com/ing-bank/lion/pull/1138/files#diff-40ce41fd27505d58be4799c69e13e3a53cec2612962afd1b6f82bef58043a6ea) and I included it on `FormControlMixin`. You can see a demo in [Theming/Combobox](https://github.com/ing-bank/lion/pull/1138/files#diff-f0fdc676f6e0d85205312e9ccc0e8cd127ac64bc6e4edd466f8479c27f6e9948), there are 3 comboboxes:

1. `lion-combobox`, styled using Shadow Parts
2. `parts-none-combobox` extended from `LionCombobox` without `exposeParts`
3. `parts-all-combobox` extended from `LionCombobox` with `exposeParts`

**Reply by jorenbroekema** [reply]
I just realized today that a NoPartsMixin implementation could also be added on the base class of the extending design system.

```js
import { LitElement as _LitElement } from '@lion/core';

const NoPartsMixin = (superclass = class {}) => class extends superclass {
  static get _exposedParts() { 
    return [];
  }
};

export class LitElement extends NoPartsMixin(_LitElement) {}
```
and then in a component, opt-in to the parts where needed.

Seems pretty acceptable to me.
